src/Roslyn.Diagnostics.Analyzers/CSharp/CSharpDoNotCapturePrimaryContructorParameters.cs

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the MIT license.  See License.txt in the project root for license information.
+
+using Analyzer.Utilities;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+using Roslyn.Diagnostics.Analyzers;
+using System.Collections.Immutable;
+
+namespace Roslyn.Diagnostics.CSharp.Analyzers
+{
+    using static RoslynDiagnosticsAnalyzersResources;
+
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class DoNotCapturePrimaryConstructorParametersAnalyzer : DiagnosticAnalyzer
+    {
+        private static readonly DiagnosticDescriptor Rule = new(
+            RoslynDiagnosticIds.DoNotCapturePrimaryConstructorParametersRuleId,
+            CreateLocalizableResourceString(nameof(DoNotCapturePrimaryConstructorParametersTitle)),
+            CreateLocalizableResourceString(nameof(DoNotCapturePrimaryConstructorParametersMessage)),
+            DiagnosticCategory.RoslynDiagnosticsMaintainability,
+            DiagnosticSeverity.Error,
+            isEnabledByDefault: false,
+            description: CreateLocalizableResourceString(nameof(DoNotCapturePrimaryConstructorParametersDescription)));
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+            context.RegisterOperationAction(AnalyzeOperation, OperationKind.ParameterReference);
+        }
+
+        private static void AnalyzeOperation(OperationAnalysisContext context)
+        {
+            var operation = (IParameterReferenceOperation)context.Operation;
+
+            if (operation.Parameter.ContainingSymbol == context.ContainingSymbol || operation.Parameter.ContainingSymbol is not IMethodSymbol { MethodKind: MethodKind.Constructor })
```

Comment:
`operation.Parameter.ContainingSymbol == context.ContainingSymbol`

Having that be the way to detect primary constructors / symbols feels fairly indirect. Was surprised there wasn't a more declarative way. Did we consider one during impl?

Reply:
We specifically didn't expose a way to know what the primary constructor is, but this actually feels more correct to me. The goal here is to understand when a parameter is being used outside of the current method, which is exactly what this expresses.